### ISSUE TITLE
episode number fallback strategy causes confusion

### ISSUE THREAD

**Issue by eprenamer** [tvrenamer/tvrenamer#168]
We face the reality that there is not necessarily a definitive numbering scheme for the episodes.  The provider we currently use may prevent several options for a given episode (and, for what it's worth, there are still frequent arguments on their forums about what's the "correct" number).

Our code has taken the approach of "prefer the 'DVD episode number', but when it's not present, fall back on the 'regular' number":

https://github.com/tvrenamer/tvrenamer/blob/eb15c842edca27a6dd6cad4c2693727338e8a0b0/src/main/com/google/code/tvrenamer/controller/TheTVDBProvider.java#L139-L166

Any such "fallback" strategy is kind of a problem, because the whole reason there are two numbering schemes is that they conflict with each other.  So if you fall back, you are mixing and matching between the two schemes.

Moreover, the "DVD episode number" corresponds to the "DVD season", which is not necessarily the same as the "Season".  And, not all episodes get released to DVD (or, not all episodes have been released yet).

Currently, we haven't exposed any of this to the user.  They have no control over which numbering scheme is used.

I am going to resolve this problem by always using the "regular" episode number.  I will open an issue for ways to better integrate the DVD numbers.

**Comment by vipuldelwadia** [comment]
Using the DVD ordering / numbering was chosen because of the problem with Firefly, where the aired order is usually not what people want.

I agree that we shouldn't really mix aired numbering with DVD numbering within a show, however if we do change it then a lot of people will have their Fireflies reordered (along with other shows I'm sure) so we should consider how best to handle this before switching to the "regular" episode number.

**Comment by eprenamer** [comment]
Thanks.  I did see the testing of Firefly, and noticed how those tests began to fail when I tried preferring the regular number.

I created a separate issue for how to handle it.  But, bluntly (sorry), this issue isn't about a feature, it's about a bug.  For many shows, the current ordering is simply wrong.  We are not getting the DVD ordering.  We're getting a complete jumble.

I'm not sure why you say the Fireflies would be "reordered".  An episode with S01E01 in the title would still be considered S01E01.  It would just have the wrong title and air date associated with it.  And that sucks, but right now, it's happening all the time.  The current scheme ensures that some epiodes will be wrong, no matter which option you prefer.  If you want the DVD ordering, some episodes are wrong, and if you want the regular ordering, some episodes are wrong.

**Comment by vipuldelwadia** [comment]
I hadn't realised that lots of shows were affected by this - in that case this is definitely a bug we need to fix. Do you have an example show that I can look at?

When I say "reordered" in the context of Firefly, what I mean is that for example, [DVD ordering](http://thetvdb.com/index.php?lid=7&order=aired&order=aired&order=dvd&seasonid=15791&seriesid=78874&tab=season&order=dvd) gives `S01E01` the title `Serenity`, however [aired order](http://thetvdb.com/index.php?lid=7&order=aired&order=aired&order=dvd&order=dvd&seasonid=15791&seriesid=78874&tab=season&order=aired) has `S01E01` as `The Train Job`. If we change the ordering to aired, then people's `S01E01 Serenity` will be renamed to `S01E01 The Train Job`, even though that episode is definitely `Serenity`.

Perhaps we need to handle this in the app by allowing the user to pick the order, like TheTVDB does on their site.

**Comment by eprenamer** [comment]
That would be great!  Would you have time to add that to the UI?

**Comment by eprenamer** [comment]
In any case, I'm almost done with a change that would keep the spirit of the original: prefer the DVD numbering, but use "regular" if DVD is not sufficient; but it will do it on a per-show basis, not per-episode.

The Show keeps track of the episodes in two ways: it has a list of every episode that the provider told us about, and then it indexes a subset (not necessarily a proper subset) of them by season and episode.  I will index them by whatever seems better based on the information downloaded.

If we add the option to the UI, we can change it in real time.  We might order by DVD initially, and if the user changes the setting, we just keep the list of episodes, clear out the index, and rebuild the index based on the new preference (and then propagate the change so that the "rename" column reflects the preference).

**Comment by eprenamer** [comment]
Well, I did misdiagnose some of the problems I was seeing.  The problems I describe are likely restricted to shows that have had only partial DVD releases.  It still comes up, but not as much as I thought.

What I was seeing was actually fanout from combining the DVD episode number, with the "regular" season number.

For example, the Futurama episode ""Proposition Infinity" was S05E04 on DVD, but S06E04 in "regular".  In this example, the actual "episode" number is the same: 4, in both cases.  But even though we thought we were preferring the DVD numbering, we'd actually use the regular season, so we'd consider this S06E04.

However, the episode "Benderama", which is S06E15 in the regular numbering, is S06E04 in the DVD numbering.  In this case, the season is the same in both cases, 6, but the episode numbers differ, and we preferred the DVD, so this would become S06E04.

So, we were mapping "Proposition Infinity" and "Benderama" to the same episode.  If we processed "Futurama-S06E04.avi", I really don't know which episode title it would prefer.  We process the episode XML in serial, so it would be whichever occurred later in the file.

**Comment by eprenamer** [comment]
This was primarily resolved in PR #170 , and modified somewhat in PR #185 (this commit:

https://github.com/tvrenamer/tvrenamer/pull/185/commits/258af48db5568229087f6fd705ec9bfd331596be
)

Now for a given show, we always use a single, consistent number scheme.  We still prefer the DVD ordering, in which case we make sure to use both season # and episode # from the DVD ordering, and include only those episodes that have DVD info.  If not enough episodes do, we choose the production ordering -- again, consistently, for every episode of the show.

**Comment by eprenamer** [comment]
I'm re-opening this.

**Comment by eprenamer** [comment]
As I mentioned, the biggest problem was that when we used the DVD episode number, we still used the "regular" season number.  So, for example, for Futurama episode "The Why of Fry", TVDB gives us "regular" season 5, episode 8, but "DVD" season 4, episode 10.0.  We turned this into season 5, episode 10.  That was the main bug.

We can continue to always prefer the DVD numbering, and falling back on "regular", as long as we take both season and episode from the same source.

With that change there is still a clear *possibility* of a bug, and I think it might exist, but I don't have an example handy.  Theoretically, there could be a show with episodes like this:
- Episode A: "regular" season 4, episode 14; "DVD" season 5, episode 1
- Episode B: "regular" season 5, episode 1; no DVD information

In that case, they'd both be mapped to S05E01, which would be problematic.

But, again, I don't have a specific example of this, and I do expect to fix the (theoretical) problem eventually by doing the numbering on a per-show basis.  But the simple fix for the primary bug is just to make sure to use season and episode together.
